src/cswinrt/code_writers.h

```diff
@@ -3549,6 +3550,551 @@ private % AsInternal(InterfaceTag<%> _) => % ?? Make_%();
+R"([global::System.ComponentModel.EditorBrowsable(global::System.ComponentModel.EditorBrowsableState.Never)]
+public static unsafe class %Marshaller
+{
+)", name);
+        
+        if (!is_type_blittable(type))
+        {
+            write_convert_to_unmanaged_method_struct(w, type);
+            write_convert_to_managed_method_struct(w, type);
+            write_dipose_method_struct(w, type);
+        }
+
+        w.write(
+R"(public static WindowsRuntimeObjectReferenceValue BoxToUnmanaged(%? value)
+{
+    return WindowsRuntimeValueTypeMarshaller.BoxToUnmanaged(value, in %ReferenceImpl.IID);
+}
+)", projection_name, name);
+
+        if (!is_type_blittable(type))
+        {
+            w.write(
+R"(public static %? UnboxToManaged(void* value)
+{
+    %? abi = WindowsRuntimeValueTypeMarshaller.UnboxToManaged<%>(value);
+    return abi.HasValue ? ConvertToManaged(abi.GetValueOrDefault()) : null;
+}
+}
+)", projection_name, abi_name, abi_name);
+        }
+        else
+        {
+            w.write(
+ R"(public static %? UnboxToManaged(void* value)
+{
+    return WindowsRuntimeValueTypeMarshaller.UnboxToManaged<%>(value);
+}
+}
+)", projection_name, projection_name);
+        }
+    }
+
+    void write_reference_impl_struct(writer& w, TypeDef const& type)
+    {
+        auto name = w.write_temp("%", bind<write_type_name>(type, typedef_name_type::ABI, false));
+        auto projection_name = w.write_temp("%", bind<write_projection_type>(type));
+        auto abi_name = w.write_temp("%", bind<write_abi_type>(type));
+        auto result_param = is_type_blittable(type) ? projection_name : abi_name;
+        std::string guid_sig = w.write_temp("%", bind<write_guid_signature>(type));
+        std::string ireference_guid_sig = "pinterface({61c17706-2d65-11e0-9ae8-d48564015472};" + guid_sig + ")";
+
+        w.write(
+R"(file static unsafe class %ReferenceImpl
+{
+    [FixedAddressValueType]
+    private static readonly ReferenceVftbl Vftbl;
+
+    private const int S_OK = unchecked((int)0x00000000);
+    private const int E_POINTER = unchecked((int)0x80004003);
+
+    static %ReferenceImpl()
+    {
+        *(IInspectableVftbl*)Unsafe.AsPointer(ref Vftbl) = *(IInspectableVftbl*)IInspectableImpl.Vtable;
+        Vftbl.get_Value = &get_Value;
+    }
+
+    public static nint Vtable
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        get => (nint)Unsafe.AsPointer(in Vftbl);
+    }
+
+    [UnmanagedCallersOnly(CallConvs = [typeof(CallConvMemberFunction)])]
+    public static int get_Value(void* thisPtr, void* result)
+    {
+        if (result is null)
+        {
+            return E_POINTER;
+        }
+
+        try
+        {)", name, name);
+        
+        if (is_type_blittable(type))
+        {
+            w.write(
+R"(            
+            var value = (%)(ComInterfaceDispatch.GetInstance<object>((ComInterfaceDispatch*)thisPtr));
+            *(%*)result = value; // Explicit cast back to the correct type)"
```

Comment:
I would not put the comment in generated code, but outside in codewriters if desired.